syntax = "proto3";
package angzarr;
option go_package = "github.com/angzarr/angzarr/proto/angzarr";
option java_package = "dev.angzarr";
option java_multiple_files = true;
option csharp_namespace = "Angzarr";
option ruby_package = "Angzarr";

import "google/protobuf/any.proto";

// CloudEvent represents a single event for external consumption.
//
// Client projectors create these by filtering/transforming internal events.
// Framework fills envelope fields (id, source, time) from Cover/EventPage
// if not explicitly set by the client.
//
// The `data` field is a protobuf Any that framework converts to JSON via
// prost-reflect using the descriptor pool. Clients should pack a "public"
// proto message that omits sensitive fields.
message CloudEvent {
  // Event type (e.g., "com.example.order.created").
  // Default: proto type_url suffix from original event.
  string type = 1;

  // Event payload as proto Any.
  // Framework converts to JSON for CloudEvents output.
  // Client should filter sensitive fields before packing.
  google.protobuf.Any data = 2;

  // Custom extension attributes.
  // Keys should follow CloudEvents naming (lowercase, no dots).
  // Framework adds correlationid automatically if present in Cover.
  map<string, string> extensions = 3;

  // Optional overrides. Framework uses Cover/EventPage values if not set.
  optional string id = 4;      // Default: {domain}:{root_id}:{sequence}
  optional string source = 5;  // Default: angzarr/{domain}
  optional string subject = 6; // Default: aggregate root ID
}

// CloudEventsResponse is returned by client projectors in Projection.projection.
//
// Framework detects this type by checking projection.type_url and routes
// the events to configured sinks (HTTP webhook, Kafka).
//
// Client may return 0 events (skip), 1 event (typical), or N events
// (fan-out scenarios like multi-tenant notifications).
message CloudEventsResponse {
  repeated CloudEvent events = 1;
}
